pkg/jaegerquery/trace_query.go

```diff
@@ -0,0 +1,450 @@
+		}
+	}
+	return nil
+}
+
+func populateEvents(
+	spanEventSlice pdata.SpanEventSlice,
+	dbResult *spanDBResult) {
+
+	n := len(*dbResult.eventNames)
+	for i := 0; i < n; i++ {
+		event := spanEventSlice.AppendEmpty()
+		event.SetName((*dbResult.eventNames)[i])
+		event.SetTimestamp(pdata.NewTimestampFromTime((*dbResult.eventTimes)[i]))
+		event.SetDroppedAttributesCount(uint32((*dbResult.eventDroppedTagsCount)[i]))
+		event.Attributes().InitFromMap(makeAttributes((*dbResult.eventTags)[i]))
+	}
+}
+
+func populateLinks(
+	spanEventSlice pdata.SpanLinkSlice,
+	dbResult *spanDBResult) error {
+
+	n := len(*dbResult.linksLinkedSpanIds)
+
+	var linkedTraceIds [][16]byte
+	if err := dbResult.linksLinkedTraceIds.AssignTo(&linkedTraceIds); err != nil {
+		return fmt.Errorf("linksLinkedTraceIds: AssignTo: %w", err)
+	}
+
+	for i := 0; i < n; i++ {
+		link := spanEventSlice.AppendEmpty()
+
+		link.SetTraceID(pdata.NewTraceID(linkedTraceIds[i]))
+
+		spanId := makeSpanId(&(*dbResult.linksLinkedSpanIds)[i])
+		link.SetSpanID(spanId)
+
+		if (*dbResult.linksTraceStates)[i] != nil {
+			traceState := *((*dbResult.linksTraceStates)[i])
+			link.SetTraceState(pdata.TraceState(traceState))
+		}
+		link.SetDroppedAttributesCount(uint32((*dbResult.linksDroppedTagsCount)[i]))
+		link.Attributes().InitFromMap(makeAttributes((*dbResult.linksTags)[i]))
+	}
+	return nil
+}
+
+// makeAttributes makes attribute map using tags.
+func makeAttributes(tags map[string]interface{}) map[string]pdata.AttributeValue {
+	m := make(map[string]pdata.AttributeValue, len(tags))
+	// todo: attribute val as array?
+	for k, v := range tags {
+		switch val := v.(type) {
+		case int64:
+			m[k] = pdata.NewAttributeValueInt(val)
+		case bool:
+			m[k] = pdata.NewAttributeValueBool(val)
+		case string:
+			m[k] = pdata.NewAttributeValueString(val)
+		case float64:
+			m[k] = pdata.NewAttributeValueDouble(val)
+		case []byte:
+			m[k] = pdata.NewAttributeValueBytes(val)
+		default:
+			panic(fmt.Sprintf("unknown tag type %T", v))
+		}
+	}
+	return m
+}
+
+func makeTraceId(s pgtype.UUID) (pdata.TraceID, error) {
+	var bSlice [16]byte
+	if err := s.AssignTo(&bSlice); err != nil {
+		return pdata.TraceID{}, fmt.Errorf("trace id assign to: %w", err)
+	}
+	return pdata.NewTraceID(bSlice), nil
+}
+
+func makeSpanId(s *int64) pdata.SpanID {
+	if s == nil {
+		// Send an empty Span ID.
+		return pdata.NewSpanID([8]byte{})
+	}
+
+	b8 := ingestor.Int64ToByteArray(*s)
+	return pdata.NewSpanID(b8)
+}
+
+func makeKind(s string) pdata.SpanKind {
+	switch s {
+	case "SPAN_KIND_CLIENT":
+		return pdata.SpanKindClient
+	case "SPAN_KIND_SERVER":
+		return pdata.SpanKindServer
+	case "SPAN_KIND_INTERNAL":
+		return pdata.SpanKindInternal
+	case "SPAN_KIND_CONSUMER":
+		return pdata.SpanKindConsumer
+	case "SPAN_KIND_PRODUCER":
+		return pdata.SpanKindProducer
+	case "SPAN_KIND_UNSPECIFIED":
+		return pdata.SpanKindUnspecified
+	default:
+		panic(fmt.Sprintf("unknown span kind: %s", s))
```

Comment:
I don't think this should be panic, as `s` will be read from db, so an error makes more sense than crashing entire promscale.